### PR TITLE
Create a new escalationCase record when primary case on an escalation…

### PR DESCRIPTION
**Pull request by preeti-preeti**
… is not null

### CODE REVIEW

File: src/classes/EscalationTrigger.cls [hunk 1]
```diff
@@ -112,4 +140,16 @@ public with sharing class EscalationTrigger extends BaseTrigger {
...
 	}
+
+	/**
+	* Perform actions that should be done in the 'after' part of a trigger
+	*/
+	@TestVisible
+	protected override void runAfter() {
+		super.runAfter();
+
+		if (this.numberOfBeforeRuns == 1 && this.numberOfAfterRuns == 1) {
```
Comment (pcon) [inline src/classes/EscalationTrigger.cls]: This should just be

```java
if (this.numberOfAfterRuns == 1) {
```

We don't care about the number of before runs in the after context

File: src/classes/EscalationTrigger.cls [hunk 2]
```diff
@@ -102,6 +102,34 @@ public with sharing class EscalationTrigger extends BaseTrigger {
...
+	@TestVisible
+	private void createEscalationCaseWhenPrimaryCaseIsNotNull() {
+		if (this.isDelete || this.isUpdate) {
+			return;
+		}
+
+		List<rh_escal__EscalationCase__c> escalCaseToInsert = new List<rh_escal__EscalationCase__c>();
+
+		for (rh_escal__Escalation__c escalation : this.newObjs) {
+			Boolean nullPrimaryCase = (escalation.rh_escal__PrimaryCase__c == null);
```
Comment (pcon) [inline src/classes/EscalationTrigger.cls]: Flip this logic a little bit

```java
for (rh_escal__Escalation__c escalation : this.newObjs) {
    if (escalation.rh_escal__PrimaryCase__c == null) {
        continue;
    }

    escalCaseToInsert.add(new rh_escal__EsclationCase__c(
        rh_escal__Case__c = escalation.rh_escal__PrimaryCase__c,
        rh_escal__Primary__c = true,
        rh_escal__Escalation__c = escalation.Id
    ));
}
```

File: src/classes/EscalationTrigger_Test.cls [hunk 3]
```diff
@@ -186,4 +186,58 @@ public class EscalationTrigger_Test {
...
+
+		Test.stopTest();
+
+		testEscalation = TestUtils.fetchEscalation(testEscalation);
+
+		rh_escal__EscalationCase__c escalCase = TestUtils.getEscalationCase(testEscalation, testCase);
+
+		System.assertEquals(testCase.Id, testEscalation.rh_escal__PrimaryCase__c, 'The primary case is not set correctly.');
+		System.assertEquals(testAccount.Id, testEscalation.rh_escal__Account__c, 'The account is not set correctly.');
+		System.assertNotEquals(null, escalCase, 'Did not get the expected escal case record.');
```
Comment (pcon) [inline src/classes/EscalationTrigger_Test.cls]: Also add some asserts to make sure that all three fields you set in the trigger are set in the `escalCase` object

```java
System.assertEquals(testCase.Id, escalCase.rh_escal__Case__c, 'The case id was not set correctly');
System.assert(escalCase.rh_escal__Primary__c, 'This should be the primary case');
System.assertEquals(testEscalation.Id, escalCase.rh_escal__Escalation__c, 'The escalation id was not set correctly');

File: src/classes/EscalationTrigger_Test.cls [hunk 4]
```diff
@@ -186,4 +186,58 @@ public class EscalationTrigger_Test {
...
+
+		Test.startTest();
+
+		insert testEscalation;
+
+		Test.stopTest();
+
+		testEscalation = TestUtils.fetchEscalation(testEscalation);
+
+		rh_escal__EscalationCase__c escalCase = TestUtils.getEscalationCase(testEscalation, testCase);
```
Comment (pcon) [inline src/classes/EscalationTrigger_Test.cls]: This doesn't do what we want it to do.  We need to add a new method to `TestUtils` that fetches all escalation cases and then we can do our asserts on that

```java
/**
* Fetches all escalation cases for a given escalation
*
* @param escalation The escalation to get it's cases
* @return The escalation cases
*/
global static List<rh_escal__EscalationCase__c> fetchEscalationCases(rh_escal__Escalation escalation) {
    return [
        select rh_escal__Case__c,
            rh_escal__Escalation__c,
            rh_escal__Primary__c
        from rh_escal__EscalationCase__c
        where rh_escal__Escalation__c = :escalation.Id
    ];
}
```

Then update line 212 to be

```java
List<rh_escal__EscalationCase__c escalCases = TestUtils.fetchEscalationCases(testEscalation);
System.assertEquals(1, escalCases.size(), 'Did not get the expected number of cases back');
rh_escal__EscalationCase__c escalCase = escalCases.get(0);
```

File: src/classes/EscalationTrigger_Test.cls [hunk 5]
```diff
@@ -186,4 +186,58 @@ public class EscalationTrigger_Test {
...
+
+		System.assertEquals(testCase.Id, testEscalation.rh_escal__PrimaryCase__c, 'The primary case is not set correctly.');
+		System.assertEquals(testAccount.Id, testEscalation.rh_escal__Account__c, 'The account is not set correctly.');
+		System.assertNotEquals(null, escalCase, 'Got the escal case record that we did not expected.');
```
Comment (pcon) [inline src/classes/EscalationTrigger_Test.cls]: Remove this since it's not actually doing what we want

File: src/classes/EscalationTrigger_Test.cls [hunk 6]
```diff
@@ -186,4 +186,58 @@ public class EscalationTrigger_Test {
...
+
+		Test.startTest();
+
+		insert testEscalation;
+
+		Test.stopTest();
+
+		testEscalation = TestUtils.fetchEscalation(testEscalation);
+
+		rh_escal__EscalationCase__c escalCase = TestUtils.getEscalationCase(testEscalation, testCase);
```
Comment (pcon) [inline src/classes/EscalationTrigger_Test.cls]: Do the same `TestUtils.fetchEscalationCases` call as above but assert that the `.size()` is 0

File: src/classes/EscalationTrigger_Test.cls [hunk 7]
```diff
@@ -186,4 +186,58 @@ public class EscalationTrigger_Test {
...
+
+		System.assertEquals(testCase.Id, testEscalation.rh_escal__PrimaryCase__c, 'The primary case is not set correctly.');
+		System.assertEquals(testAccount.Id, testEscalation.rh_escal__Account__c, 'The account is not set correctly.');
+		System.assertNotEquals(null, escalCase, 'Got the escal case record that we did not expected.');
+	}
```
Comment (pcon) [inline src/classes/EscalationTrigger_Test.cls]: Please also add a test that updating an escalation that has a null primary to case to have a primary case does not create the escalation case record.

File: src/classes/EscalationTrigger.cls [hunk 8]
```diff
@@ -102,6 +102,34 @@ public with sharing class EscalationTrigger extends BaseTrigger {
 	}
 
 	/**
+	* Auto populate the rh_escal__EscalationCase__c	record if primary case on the escalation is not null.
+	*/
+	@TestVisible
+	private void createEscalationCaseWhenPrimaryCaseIsNotNull() {
+		if (this.isDelete || this.isUpdate) {
```
Comment (pcon) [inline src/classes/EscalationTrigger.cls]: Make this

```java
if (!this.isInsert) {
```

It's unlikely that we'd add support for `undelete` or another type, but we want to be specific that this is only for inserts.

**Comment by preeti-preeti** [comment]
Patrick, I am confused with this implementation, I have some concerns like-
1. If we are inserting a new escalationCase every time  when a new escalation record is created with a primary case value then shouldn't we allow the same to happen when an **escalation is updated** to have a latest primary case every time its is updated (either from null to something or from a -> b)?
2. As you I am also not liking PrimaryCase field on an escalation if we can associate an escalation with multiple cases. And as you know the same logic is taken care of when we create a new escalationCase and if we are setting it as primary then escalation primaryCase is updated accordingly so can we make primaryCase field on escalation as **readOnly** from API as well as SFDC UI (and can hide this field from escalationEdit vf page once we are ready with a vf page of escalationEdit functionality.)

I need your opinion on this...plz suggest.

**Comment by preeti-preeti** [comment]
@pcon http://pastebin.test.redhat.com/477839 this is how it is implemented in CCM.

**Comment by pcon** [comment]
To your points

1. I would say that we should probably add some logic that doesn't allow the primary case to be edited after the escalation is inserted.  I would see this being a static var on the trigger class that can be unset when we update the escalation via the escalation case trigger
2. I agree, but this is the way to allow the majority case (single escalation single case) but still allow the minority case (single escalation multiple case).  When this was originally implemented there was no escalation case object.  It was added after the fact.  It's still useful to have a primary case field for reporting purposes since single case is still the primary vector.

**Comment by preeti-preeti** [comment]
closing this pull request, will send a new request along with the changes fields modifications
